cmake/HPX_DetectCppDialect.cmake

```diff
@@ -6,77 +6,46 @@
+  if(HPX_WITH_CXX17)
+    set(CXX_FLAG -std=c++17 PARENT_SCOPE)
+    hpx_add_config_cond_define(BOOST_NO_AUTO_PTR)
+    hpx_info("C++ mode used: C++17")
   else()
-
-    # Try -std=c++17 first
-    if(HPX_WITH_CXX17 OR NOT (("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Intel")
+    # ... otherwise try -std=c++1z
+    if(HPX_WITH_CXX1Z OR NOT (("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Intel")
                               AND (CMAKE_CXX_COMPILER_VERSION VERSION_LESS 17)))
-      check_cxx_compiler_flag(-std=c++17 HPX_WITH_CXX17)
+      check_cxx_compiler_flag(-std=c++1z HPX_WITH_CXX1Z)
     endif()
 
-    if(HPX_WITH_CXX17)
-      set(CXX_FLAG -std=c++17 PARENT_SCOPE)
+    if(HPX_WITH_CXX1Z)
+      set(CXX_FLAG -std=c++1z PARENT_SCOPE)
       hpx_add_config_cond_define(BOOST_NO_AUTO_PTR)
-      hpx_info("C++ mode used: C++17")
+      hpx_info("C++ mode used: C++1z")
     else()
-      # ... otherwise try -std=c++1z
-      if(HPX_WITH_CXX1Z OR NOT (("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Intel")
-                                AND (CMAKE_CXX_COMPILER_VERSION VERSION_LESS 17)))
-        check_cxx_compiler_flag(-std=c++1z HPX_WITH_CXX1Z)
+      # ... otherwise try -std=c++14
+      if(HPX_WITH_CXX14)
+        check_cxx_compiler_flag(-std=c++14 HPX_WITH_CXX14)
       endif()
 
-      if(HPX_WITH_CXX1Z)
-        set(CXX_FLAG -std=c++1z PARENT_SCOPE)
-        hpx_add_config_cond_define(BOOST_NO_AUTO_PTR)
-        hpx_info("C++ mode used: C++1z")
-      else()
-        # ... otherwise try -std=c++14
-        if(HPX_WITH_CXX14 OR NOT (("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Intel")
-                                  AND (CMAKE_CXX_COMPILER_VERSION VERSION_LESS 17)))
-          check_cxx_compiler_flag(-std=c++14 HPX_WITH_CXX14)
-        endif()
-
-        if(HPX_WITH_CXX14)
-          set(CXX_FLAG -std=c++14 PARENT_SCOPE)
-          # The Intel compiler doesn't appear to have a fully functional
-          # implementation of C++14 constexpr. It's fine with our C++14 constexpr
-          # usage in HPX but chokes on Boost.
-          # FIXME: This should be replaced with a version-based check in the future
-          # when the Intel compiler is able to build Boost with -std=c++14.
-          if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Intel")
-            hpx_add_config_cond_define(BOOST_NO_CXX14_CONSTEXPR)
-          endif()
-          hpx_info("C++ mode used: C++14")
-        else()
-          # ... otherwise try -std=c++1y
-          if(HPX_WITH_CXX1Y OR NOT (("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Intel")
-                                    AND (CMAKE_CXX_COMPILER_VERSION VERSION_LESS 17)))
-            check_cxx_compiler_flag(-std=c++1y HPX_WITH_CXX1Y)
-          endif()
-
-          if(HPX_WITH_CXX1Y)
-            set(CXX_FLAG -std=c++1y PARENT_SCOPE)
-            hpx_info("C++ mode used: C++1y")
-          else()
-            # ... otherwise try -std=c++11
-            check_cxx_compiler_flag(-std=c++11 HPX_WITH_CXX11)
-            if(HPX_WITH_CXX11)
```

Comment:
I think we should warn if somebody uses `HPX_WITH_CXX11` (or similar), for instance in build scripts, so that people are made aware of this (breaking) change.

Reply:
Yep, that's a good point. We may have to hold off with this PR and just give a warning in the next release... and then a hard error after that.

Reply:
Added a hard error for `HPX_WITH_CXX11` and co. We should definitely wait with this PR until after the release and add a warning for this release.

I'd argue that we should actually deprecate the `HPX_WITH_CXX` options and set `CXX_STANDARD` instead (or have the user set that).

Reply:
Ok, I'd be fine with this change. Those would have to be phased out in the usual way, however. People have build scripts they use every day that possibly have the `HPX_WITH_CXX` switches, removing support for those needs to be communicated prominently.

Reply:
Absolutely. It's a separate issue though and I'll have a look at it in a separate PR.